### PR TITLE
Implement Tanh Gelu Approximation

### PR DESCRIPTION
**Pull request by rdspring1**
1. Implements #39853 
2. Adds approximate boolean flag to Gelu
3. Enables Tanh Gelu approximation
4. Adds double backward support for Gelu
5. Enable Tanh Gelu in NvFuser 

```
def gelu(x, approximate : str = 'none'):
    if approximate == 'tanh':
        # sqrt(2/pi) = 0.7978845608028654
        return 0.5 * x * (1.0 + torch.tanh(0.7978845608028654 * (x + 0.044715 * torch.pow(x, 3.0))))
    else:
        return x * normcdf(x)
```

Linking XLA PR - https://github.com/pytorch/xla/pull/3039

### CODE REVIEW

**Comment by vadimkantorov** [comment]
From what I understood, there are now multiple approximations in use in third-party frameworks. Should the arg then be called `approx_tanh`/`tanh_approx`/`approx='tanh'`?

**Comment by vadimkantorov** [comment]
> From what I understood, there are now multiple approximations in use in third-party frameworks. Should the arg then be called `approx_tanh`/`tanh_approx`/`approx='tanh'`? (extensible if the other approx variant gets supported in the future)

@rwightman @hendrycks Could you please comment on this?

**Comment by hendrycks** [comment]
Possibilities:
It could be approx=True, which would then default to a sigmoid approximation.
approx=True -> sigmoid approximation

Or there could be both the sigmoid (for lowest memory usage) and tanh approximation (for cross-compatibility with tensorflow).
approx='tanh', approx='sigmoid', approx=True (equivalent to approx='sigmoid')

**Comment by jjsjann123** [comment]
FYI, backward_compatibility_check_test is just a warning that you are changing the signature and need to add the function to an allow list.

**Comment by rwightman** [comment]
@vadimkantorov @hendrycks re the arg, if it is going to be an arg I think the string approach is good as there are different users/use cases of both approx out there, probably don't need to bool + string overload...

However, personally I don't see much point in squishing it into one fn. All 3 (original, sigmoid approx, tanh approx) aren't numerically compatible with each other (ie you can't just change the flag and expect a trained network to work the same, you have to fine-tune). I think it's more clear in that respect if they are separate activation instances. Having it as an arg isn't going to save one code/effort as it'll require a partial wrap anyways, might as well make it `gelu`, `gelu_approx_sigmoid`, `gelu_approx_tanh` ....

**Comment by hendrycks** [comment]
> However, personally I don't see much point in squishing it into one fn.

Tensorflow/keras put it in one function. In other modules you can customize them to get slightly different behavior (e.g., nn.Linear(..., bias=False), torch.svd_lowrank(A, q=6, niter=2)). The other advantage is that it doesn't clutter the PyTorch namespace as much.

> probably don't need both the bool + string overload

Yes, so it could be approx='sigmoid', approx='tanh' (and maybe approx='True' depending on what others think).

File: torch/nn/modules/activation.py
Comment (jjsjann123) [inline torch/nn/modules/activation.py:674]: return `None` on __init__

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:520]: clang tidy on tab

File: test/cpp/jit/test_gpu.cpp
Comment (jjsjann123) [inline test/cpp/jit/test_gpu.cpp]: Note. we would not want to break onnx export.
Would is be safe to convert it during onnx export and spit out vanilla gelu without the approximation?

Comment (jjsjann123) [reply test/cpp/jit/test_gpu.cpp]: NVM... sorry I was looking at the wrong code.

**Comment by hendrycks** [comment]
Any update on this?

**Review by eellison — COMMENTED** [review]
is this intended to no longer be in draft ?

**Comment by rdspring1** [comment]
I was meaning to run some performance benchmarks marking the PR for review.

File: torch/nn/modules/activation.py
Comment (vadimkantorov) [inline torch/nn/modules/activation.py:650]: It should be good to add references to models / codebases that use this particular approximation, e.g. refer to original google bert impl (taking account that there are two different approximations in the wild)

**Comment by rdspring1** [comment]
I don't have access to a cpu with bfloat16 support, so here is the benchmark script.
```
import torch
import time

ITER = 500

sizes = [(2, 8),
(4, 16),
(16, 20),
(32, 20),
(64, 14),
(80, 12),
(1, 1024),
(2, 1024),
(4, 1024),
(8, 1024),
(16, 1024),
(32, 1024),
(64, 1024),
(128, 1024),
(256, 1024),
(8, 384, 3072)]

print("dtype; device; approximate; input-size; time-per-iteration (ms)")
for dtype in [torch.bfloat16, torch.float32]:
  for device in ["cpu", "cuda"]:
    for approximate in [False, True]:
      for size in sizes:
        x = torch.randn(size, dtype=dtype, device=device)
        total_time = 0
        for i in range(ITER):
          torch.cuda.synchronize()
          start = time.time()
          o = torch.nn.functional.gelu(x, approximate)
          torch.cuda.synchronize()
          end = time.time()
          total_time += end - start
        print("{}; {}; {}; {}; {:.6f}".format(dtype, device, approximate, size, total_time / ITER * 1000))
      print("\n\n")
```

**Comment by hendrycks** [comment]
Would it be easy to take the SiLU code to also add x * sigmoid(1.702 * x) as an approximation?
This cheaper popular approximation was discussed in https://github.com/pytorch/pytorch/issues/39853 and earlier in the thread.

**Comment by ngimel** [comment]
cc @jbschlosser for added option to the API.

**Comment by ngimel** [comment]
The benchmarking graphs are pretty uninformative, most of the graph is flat line indicating same time per iteration. Can you show achieved bandwidth and the sizes where it really matters?

**Review by jbschlosser — COMMENTED** [review]
> cc @jbschlosser for added option to the API.

Based on the discussion here, it seems possible that we'd want both the tanh and sigmoid approximations in the future, so my opinion here is to to prefer the string-based API (e.g. `approx='tanh'`) to allow for this (scope of this PR should still be tanh approx only). Default should be `approx=None` for BC.

I realize this differs from [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/nn/gelu) and [JAX](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.gelu.html#jax.nn.gelu), which is somewhat unfortunate. I don't think it's worth supporting `approx=True` since it would be ambiguous which approximation is used.

Note that for JIT FC reasons, the new kwarg should be added to the ATen operator `gelu` only but no changes should be made yet to `torch/nn/modules/activation.py` or `torch/nn/functional.py` to use it. After the 2 week FC window has passed, these files can be changed to use the kwarg. See #67565 for more details.

**Comment by jjsjann123** [comment]
With the update on gelu signature we are now failing onnx tests.
Wondering if someone could point us to a good place to update the onnx export so we can make an exception to shoehorn aten::gelu into onnx IR?

**Comment by dzhulgakov** [comment]
You probably need to updates one of the torch/onnx/symbolic files. @BowenBao can help if needed

**Comment by jjsjann123** [comment]
Ryan has got onnx things working and followed the FC workaround from #67565
Can we get someone to review and landing this PR?

**Comment by hendrycks** [comment]
Any updates?

**Comment by jjsjann123** [comment]
I've been periodically bumping upstream for further review/stamp, but we haven't got much updates lately :cry:

**Review by jbschlosser — APPROVED** [review]
Sorry for the long stamp delay - LGTM!

**Comment by jbschlosser** [comment]
Do you mind doing another quick rebase - having some trouble getting this landed due to the merge conflicts

**Comment by rdspring1** [comment]
I pulled the latest changes and resolved the merge conflicts.

**Comment by jbschlosser** [comment]
FYI I'm getting several internal test failures - investigating now.

**Comment by jjsjann123** [comment]
> FYI I'm getting several internal test failures - investigating now.

Thanks for helping out. Internal test failures are tricky to handle due to the lack of visibility, but feel free to reach out if you think we'll be of help.

**Comment by jbschlosser** [comment]
Update: I don't have bandwidth now to get the internal issues resolved, but @eellison mentioned he can take care of it Friday.

**Comment by eellison** [comment]
@rdspring1 can you rebase please ?

**Comment by rdspring1** [comment]
I merged the latest changes.

**Comment by jjsjann123** [comment]
Bumping for updates~

**Comment by anjali411** [comment]
This PR seems to be breaking onnx and jit tests:
```
======================================================================
ERROR [0.056s]: test_variant_consistency_jit_nn_functional_gelu_cpu_float32 (__main__.TestJitCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 753, in test_wrapper
    return test(*args, **kwargs)
  File "test_ops.py", line 1024, in test_variant_consistency_jit
    no_grad=not _requires_grad, no_gradgrad=not op.supports_gradgrad)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_jit.py", line 91, in check_against_reference
    outputs_test = self.runAndSaveRNG(func, nograd_inputs, kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_jit.py", line 160, in runAndSaveRNG
    results = func(*inputs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/jit_metaprogramming_utils.py", line 339, in script_fn
    self.assertExportImport(fn.graph, tensors)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_jit.py", line 148, in assertExportImport
    self.assertExportImportModule(m, inputs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_jit.py", line 152, in assertExportImportModule
    a = self.runAndSaveRNG(m, inputs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_jit.py", line 160, in runAndSaveRNG
    results = func(*inputs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 427, in prof_func_call
    return prof_callable(func_call, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 424, in prof_callable
    return callable(*args, **kwargs)
torch.jit.Error: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "<string>", line 3, in forward

def the_method(i0):
    return torch.nn.functional.gelu(i0, approximate='tanh')
           ~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 1679, in gelu
        return torch._C._nn.gelu(input)
    else:
        raise RuntimeError("TorchScript currently does not support approximate in nn.Gelu")
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: TorchScript currently does not support approximate in nn.Gelu

```

**Comment by jbschlosser** [comment]
@rdspring1 Do you mind rebasing again? Sorry for the trouble - I've been working hard on this and some internal model BC issues have been holding it up but I've got a path forward now.

**Comment by rdspring1** [comment]
Thanks, Joel, for your help! I pulled the latest changes. 

`TestJitCPU.test_variant_consistency_jit_nn_functional_gelu_cpu_float32` passes on my local branch.

**Comment by mruberry** [comment]
This was reverted for breaking mobile builds, but please add me as a reviewer before attempting to land it as there are UX issues with this PR

**Review by mruberry — CHANGES_REQUESTED** [review]
Needs UX and BC/FC review.

**Comment by jbschlosser** [comment]
Regarding BC / FC: this breaks old internal mobile models. Because they still have bytecode v4, the BC workaround for dealing with new default-valued kwargs does not apply (requires at least bytecode v5). As an alternative to re-serializing these old models (which requires approx. 5 different production teams to do so), there's the possibility of using mobile upgraders. Unfortunately, mobile upgraders are still in development and haven't worked for me yet due to some design details. I'd very much like to get this PR in for 1.11, but we need to solve these mobile issues.

File: torch/nn/functional.py
Comment (jbschlosser) [inline torch/nn/functional.py:1655]: Please make this new arg kw-only: `def gelu(input: Tensor, *, approximate: str = 'none') -> Tensor`. Similar changes may be needed throughout

File: torch/nn/functional.py
Comment (jbschlosser) [inline torch/nn/functional.py:1670]: I think this line should be removed, as the call is done below

**Review by jbschlosser — COMMENTED** [review]
@rdspring1 Apologies for the pain involved here - this PR has unfortunately unearthed a fun BC issue with a difficult solution. We're working internally to get this solved, but no complete solution exists yet.

In the meantime, do you mind making a couple small tweaks to the PR to make the new arg kw-only & I will update you as we identify a working solution. Thanks for your dedication to keeping this updated & we will get it in eventually!

File: torch/nn/functional.py
Comment (rdspring1) [reply torch/nn/functional.py]: I'm seeing this torchscript error:
```
torch.jit.frontend.NotSupportedError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults:
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 1655
def gelu(input: Tensor, *, approximate: str = 'none') -> Tensor:
                                              ~~~~~ <--- HERE
    r"""gelu(input, approximate = 'none') -> Tensor
```

**Comment by jbschlosser** [comment]
@atalman Why was this closed? We are still working on it.

**Comment by jjsjann123** [comment]
> Regarding BC / FC: this breaks old internal mobile models. Because they still have bytecode v4, the BC workaround for dealing with new default-valued kwargs does not apply (requires at least bytecode v6). As an alternative to re-serializing these old models (which requires approx. 5 different production teams to do so), there's the possibility of using mobile upgraders. Unfortunately, mobile upgraders are still in development and haven't worked for me yet due to some design details. I'd very much like to get this PR in for 1.11, but we need to solve these mobile issues.

Do we roughly know when did we bump bytecode to v6? We have internal team that's also running into this issue.

**Comment by jbschlosser** [comment]
> Do we roughly know when did we bump bytecode to v6? We have internal team that's also running into this issue.

@jjsjann123 I believe it was bumped in #58852.

File: torch/nn/functional.py
Comment (jbschlosser) [reply torch/nn/functional.py]: That's no good. We'll have to leave this as non-kw-only, sadly.

Comment (mruberry) [reply torch/nn/functional.py:1685]: We definitely should!

File: aten/src/ATen/native/native_functions.yaml
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:3731]: I propose we change the type of `approximate` to be `str` for all overloads in `native_functions.yaml` (note: C++-side, `str` becomes `c10::string_view`).
* This removes the str / int discrepancy between python / C++ and allows us to remove the `torch.nn.functional.gelu` function wrapper. In `torch.nn.functional.py`, the gelu code should once again use the approach of:
```python
gelu = _add_docstr(
    torch._C._nn.gelu,
    r"""
    ...
    """
```
* Now we no longer need the `elif not torch.jit.is_scripting` nonsense because #67565 no longer applies.
* We should be able to make the new arg kw-only again (e.g. `gelu.out(Tensor self, *, str approximate='none', Tensor(a!) out) -> Tensor(a!)`). The test error comes from the *wrapper* accepting kw-only args, so no wrapper -> no problem.
* There's no longer a need for the new `_gelu.py` file.
* Can still keep an enum defined for nice C++-side usage (see what's done for [`scatter.reduce`](https://github.com/pytorch/pytorch/blob/2d4513e4c43aa11860785bdf73c0921588b7fb11/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L77-L85)).

**Review by jbschlosser — COMMENTED** [review]
Hey @rdspring1, some good news finally: I've got a working solution for our internal issues that should allow us to land this PR. Added a comment below to address the current failing OSS checks, then we should be good to go.

**Comment by rwightman** [comment]
After all the breaks, and open/close (I'm on follow due to earlier comment), still seems like having separate acts would have been cleaner ;) ... hardsigmoid isn't combined with sigmoid, hardswish with silu, in practical use you're going to have to bind the approx arg with partial, wrap in factory / helper fns.. it doesn't appear to make life much easier vs `gelu` `gelu_sigmoid`, `gelu_tanh`....

**Comment by jbschlosser** [comment]
> After all the breaks, and open/close (I'm on follow due to earlier comment), still seems like having separate acts would have been cleaner ;) ... hardsigmoid isn't combined with sigmoid, hardswish with silu, in practical use you're going to have to bind the approx arg with partial, wrap in factory / helper fns.. it doesn't appear to make life much easier vs `gelu` `gelu_sigmoid`, `gelu_tanh`....

Well it certainly wasn't easy going the kwarg direction (due to awful internal processes that arguably should be addressed regardless), but I like that it fits with what both JAX and TF do. Binding the approximate arg is essentially what using the module form `nn.GELU` gets you, so I don't see too much of a usability problem, but I'm open to being convinced otherwise.

**Comment by rwightman** [comment]
> > After all the breaks, and open/close (I'm on follow due to earlier comment), still seems like having separate acts would have been cleaner ;) ... hardsigmoid isn't combined with sigmoid, hardswish with silu, in practical use you're going to have to bind the approx arg with partial, wrap in factory / helper fns.. it doesn't appear to make life much easier vs `gelu` `gelu_sigmoid`, `gelu_tanh`....
> 
> Well it certainly wasn't easy going the kwarg direction (due to awful internal processes that arguably should be addressed regardless), but I like that it fits with what both JAX and TF do. Binding the approximate arg is essentially what using the module form `nn.GELU` gets you, so I don't see too much of a usability problem, but I'm open to being convinced otherwise.

It's not a huge usability problem either way. More in light of the mess that the arg route seems to have involved (and having run into numerous past torchscript regressions omissions when args were added in other places).

In my typical use cases there is slightly more code to deal with the arg form as I want to pass act fn as configurable callables (that can be instantiated without any args for activations) down through model hierarchy via arguments or config classes. So the `cfg.act_layer = partial(nn.GELU, approximate='tanh')` is marginally messier, requiring an extra import for functools or my own wrapping classes/fns, than `cfg.act_layer = nn.GELUApproxTanh`

File: aten/src/ATen/native/native_functions.yaml
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:3742]: This should fix the failing checks:
```
gelu(Tensor self, *, str approximate='none') -> Tensor
```

File: tools/autograd/derivatives.yaml
Comment (jbschlosser) [inline tools/autograd/derivatives.yaml:1817]: This should fix the failing checks:
```
gelu_backward(Tensor grad_output, Tensor self, *, str approximate='none') -> Tensor
```

**Review by jbschlosser — COMMENTED** [review]
@rdspring1 Did a quick pass - placing the `*` directly before the new kwarg in all signatures in `native_functions.yaml` / `derivatives.yaml` should fix the failing checks. Once that's done, I'll merge this thing

**Comment by rdspring1** [comment]
gradcheck currently doesn't accept kwargs. Will that be an issue?
https://github.com/pytorch/pytorch/blob/0fa24f1be879a198fe4620d73cc353542c09b70f/test/test_nn.py#L9176

I'm working on string argument support for NvFuser / NNC.

**Comment by ngimel** [comment]
@rdspring1 the way of checking functions that have kwargs with gradcheck is to create a partial that has all the kwargs set.

**Comment by ngimel** [comment]
Given that gelu is already tested in OpInfo, please remove test_gelu from test_nn (it's not doing any testing that's not done in OpInfo already), and add an entry to OpInfo that would test with a different kwarg, and add corresponding `ref` implementations for the different kwargs.

File: test/test_jit_cuda_fuser.py
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:2195]: Does this actually pass the test?!

We disabled fusion guard in CI tests, that means when you change `mode`, it's running the same graph and ignores the `tanh` passed in the second time.

You might want to turn on guard mode again to run both mode in a single test like this one: https://github.com/pytorch/pytorch/blob/c2cd1d3219486dbf07662bff2a3874664c8f6978/test/test_jit_cuda_fuser.py#L974

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:67]: nitpick. I should have cleaned this and move these Symbol to somewhere maybe utils.h and reuse that across graph_fuser.cpp e.t.c.

File: torch/csrc/jit/codegen/cuda/parser.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:2493]: Can you put a note here that `string` scalars are only used in parsing rule, hence it's ok to not register it with codegen IR.

**Review by jjsjann123 — COMMENTED** [review]
LGTM, minor comments on tests

**Comment by rdspring1** [comment]
@jbschlosser I finished adding the string argument to nvfuser and tensorexpr.

File: torch/testing/_internal/common_methods_invocations.py
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:11749]: what happened here?

File: torch/testing/_internal/common_methods_invocations.py
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:4089]: sample inputs with different kwargs are good, can you also please add reference testing?

**Comment by jbschlosser** [comment]
@JackCaoG FYI this PR finally landed - do you mind landing the associated XLA PR as well? Thanks!

**Comment by JackCaoG** [comment]
@jbschlosser Sure.. I will merge that pr. Could you give me a ping before merging the pr next time? The torch/xla side pr does not get approval but I have to merge it now to unblock the master. Would be nice if I can get some heads up 😄

**Comment by jbschlosser** [comment]
@JackCaoG Sorry to rush you - will give you a heads-up next time. I've been anxious to get the PR in for 1.11 so was pushing this a bit!

**Pull request closed without merging**
